### DISCUSSION TITLE
命令式弹出式组件优化(以Modal.confirm为例)

### DISCUSSION THREAD

**Discussion by ppbl** [ant-design/ant-design#31460] [Ideas]
当前, `Modal.confirm()` 方法可以通过命令式的方式弹出一个模态框

看了下代码实现，先创建了一个 `div`, 然后把 `div` `appendChild` 到 `body` 上面，再以`div` 为容器，使用 ReactDOM.render 渲染到页面上
代码大概如下 ：
```js
  // 创建容器
  const div = document.createElement('div');
  document.body.appendChild(div);
  // 销毁组件的同时从页面移除容器节点
  function destroy() {
    const unmountResult = ReactDOM.unmountComponentAtNode(div);
    if (unmountResult && div.parentNode) {
      div.parentNode.removeChild(div);
    }
  }
  // 渲染
  render() {
    ReactDOM.render(
      <ConfirmDialog
        {...props}
      />,
      div,
    );
  }
```
[Modal.confirm 源代码](https://github.com/ant-design/ant-design/blob/master/components/modal/confirm.tsx#L30)

**当前的问题是，会在页面上创建一个额外的 `div`**

我尝试了一下, 创建 `div` 元素后即使不 `appendChild` 到 body 上面，仍可以做为 `render` 方法的容器，正确的把模态框渲染到页面（把 `div` 换成 `documentFragment` 同样），且不会产生多余的 `div`, 代码微调一下：

```js
  // 创建容器
  const container = document.createDocumentFragment()
  // 销毁组件
  function destroy() {
     ReactDOM.unmountComponentAtNode(container);
  }
  // 渲染
  render() {
    ReactDOM.render(
      <ConfirmDialog
        {...props}
      />,
      container,
    );
  }
```

这样的话由于没有往 `body` 渲染额外的 `div`，所以组件销毁的时候也不用手动 remove

**优点很明显，多余的 div 没有了，仅此而已😊**

至于缺点，表面看起来似乎没有什么缺点

我想之前这样写，是不是因为 `React.createPortal` 出现之前，由于要渲染到 body 里面必须要有一个位于body 里面的元素

现在 `<ConfirmDialog />` 这个组件本身内部使用了 `React.createPortal`，导致外部 render 方法的 container 元素的位置以及是否渲染到页面上没有那么重要了

是一个小的优化点？

ps: 其他提供静态方法弹出组件的组件类似（组件本身使用了`React.createPortal`）

**Comment by ppbl** [comment]
@afc163 @zombieJ

**Comment by afc163** [comment]
看上去可行，可以来个 PR 看看。
